wikipedia_cleanup/data_processing.py

```diff
@@ -56,45 +64,126 @@ def read_pickle_file(file_path: Path) -> List[InfoboxChange]:
+            change.property_name,
+            change.value_valid_from,
+        )
+    )
+    return changes
+
+
+def read_file_sorted(file_path: Path) -> List[InfoboxChange]:
     if "pickle" in str(file_path):
+        # pickle is already sorted
         return read_pickle_file(file_path)
     elif "json" in str(file_path):
-        return read_json_file(file_path)
-    raise ValueError("Expected a pickle or json file")
+        return sort_changes(read_json_file(file_path))
+    else:
+        raise ValueError("Expected a pickle or json file")
+
+
+def read_and_filter_file(
+    file_path_and_filters: Tuple[Path, List[AbstractDataFilter]]
+) -> Tuple[List[InfoboxChange], List[AbstractDataFilter]]:
+    file_path, filters = file_path_and_filters
+    changes = read_file_sorted(file_path)
+    return filter_changes_with(changes, filters), filters
 
 
 def get_data(
-    input_path: Path, n_files: Optional[int] = None, n_jobs: int = 0
-) -> pd.DataFrame:
+    input_path: Path,
+    n_files: Optional[int] = None,
+    n_jobs: int = 0,
+    filters: Optional[List[AbstractDataFilter]] = None,
+) -> Tuple[pd.DataFrame, List[AbstractDataFilter]]:
+    """
+    Reads the data into a pd.DataFrame from all files in parallel
+    and applies the given filters on the fly.
+    The dataframe is guaranteed to be sorted for
+    all changes of a page after the priority:
+    infobox_key, property_name, value_valid_from.
+    The returned filters contain the accumulated starts of the read.
+
+    Example usage:
+    ```
+    filters = get_default_filters()
+    data_frame = get_data(file_path, filters=filters, n_jobs=5)
+    ```
+
+    :param input_path: :pathlib.Path:
+    Path to the input folder containing decompressed jsons or pickle files.
+    :param n_files :Optional[int]:
+    Number of files to read from the input Folder. None means using all Files.
+    :param n_jobs: int:
+    Number of Jobs / Processes used for parallel reads.
+    :param filters: Optional[List[AbstractDataFilter]]:
+    (Ordered) List of Filters
+    that should be applied on the fly when loading.
+    Consider using: `get_default_filters()`
+    :return: Tuple of: All change items from all read files
+    where each all files are sorted after
+    (page_id, infobox_key, property_name, value_valid_from),
+    the resulting filters hold the accumulated stats of the read.
+    """
+    if filters is None:
+        filters = []
     files = [x for x in Path(input_path).rglob("*.output.json") if x.is_file()]
     files.extend([x for x in Path(input_path).rglob("*.pickle") if x.is_file()])
     if n_files is not None:
         n_jobs = min(n_jobs, n_files)
     n_jobs = min(n_jobs, len(files))
     files = files[slice(n_files)]
     if n_jobs > 1:
-        all_changes = process_map(read_file, files, max_workers=n_jobs)
+        all_changes, mapped_filters = zip(
+            *process_map(
+                read_and_filter_file,
+                zip(files, [filters] * len(files)),
```

Comment:
did we not find that `[filters] * len(files)` duplicates the same filters? is this not a problem here?

Reply:
No because all objects are copied anyway because of multprocessing. Basically, the object you provide are not used directly. That's why I need to return the filters in order to calculate the stats.